Issue #368 Making OCI and Docker daemon transport work on Windows

hferentschik opened the pull request:
Signed-off-by: Hardy Ferentschik <hardy@hibernate.org>

runcom commented:
I have unfortunately no way to verify this on windows. Changes make sense though

hferentschik commented on docker/daemon/daemon_src.go:
```diff
@@ -9,9 +9,18 @@ import (
-const temporaryDirectoryForBigFiles = "/var/tmp" // Do not use the system default of os.TempDir(), usually /tmp, because with systemd it could be a tmpfs.
+var temporaryDirectoryForBigFiles string
+
+func init() {
```
Instead of doing the same thing in multiple places, I was wondering whether there should be a shared util function _GetTemporaryDirectoryForBigFiles_ somewhere. AFAICT, there is no obvious location yet for these type of utility method. Should I create one? And if so where?

mtrmac replied:
Yes, please, move this into a single implementation.

Probably a subpackage in the `internal` hierarchy, something like `containers/image/internal/tmpdir`.

I wouldn’t be opposed to making this a public subpackage under `containers/image/pkg/` either, it could be useful for other software; OTOH having such a containers-independent utility under `containers/image` feels a little weird.

hferentschik replied:
Ok. I think I go for the internal/tmpdir

hferentschik commented on oci/layout/oci_transport.go:
```diff
@@ -113,7 +114,7 @@ func NewReference(dir, image string) (types.ImageReference, error) {
 	// This is necessary to prevent directory paths returned by PolicyConfigurationNamespaces
 	// from being ambiguous with values of PolicyConfigurationIdentity.
-	if strings.Contains(resolved, ":") {
+	if strings.Contains(resolved, ":") && runtime.GOOS != "windows" {
```
I am not so sure about this one. Obviously, on Windows, there will be a ':' in the path. Just adding this condition works for my use cases. I am not sure what the comment tries to say in terms of PolicyConfigurationNamespaces and PolicyConfigurationIdentity. I need some guidance here on whether a simple OS check is sufficient here or whether something else needs to happen. Maybe it can at least be a first step solution, since not allowing colon ':' in the path stops this to be working on Windows right in the beginning.

mtrmac replied:
The original comment is no longer applicable after 4eb8db31f00f5bcb25fe55341731df33f3067768 (and should be updated, along with `ValidatePolicyConfigurationIdentity`), but this restriction is still somewhat necessary, because the same ambiguity exists in `ParseReference`:

- `/path:busybox` means `dir="/path"`, `image="busybox"`
- `/path` and `/path:` means `dir="/path"`, `image=""`
- `/path:busybox:latest` means `dir="/path"`, `image="busybox:latest`";

Note that `dir="/path:busybox"`, `image="latest"` is impossible to express in the string form consumed by `ParseReference`, because it would be parsed the other way.  Giving up on `ParseReference` for this path and allowing `dir="/path:busybox"` only via `NewReference` would also be problematic because `StringWithinTransport()` must return a valid input for `ParseReference`.

---

I think it would be very reasonable to accept a `:` on Windows (perhaps, to be more strict, only in the second position, but the output of `explicitfilepath.ResolvePathToFullyExplicit` is supposed to be a valid path, and `:` is reserved on Windows, so it should not appear anywhere but after a drive letter.)

The flip side of that, though, is that `ParseReference` should likewise accept paths starting with a drive letter and a colon, and parse that as a part of the `dir` , not as the separator starting `image`.

`ValidatePolicyConfigurationIdentity` should also be updated - to treat the input only as a directory, the way `PolicyConfigurationIdentity` formats it now, and to reject `:` on non-Windows only.

hferentschik replied:
I am aware of the problems of ParseReference. In fact, I am not using ParseReference and use NewReference directly. That works fine when using the API programmatically. It is more an issue in the case of Skopeo or other potential tools which want to create a reference from a single string. 

As I mentioned before, I think the use of ':' is unfortunate in this case. I suggested back then to use a different separator like '@' or similar. 

I feel the ParseReference problems is a bigger issue which I'd like to defer to another issue. WDYT?

In moving this along, is a OS specific check ok, together with an adjustment to ValidatePolicyConfigurationIdentity?

mtrmac replied:
> I feel the ParseReference problems is a bigger issue which I'd like to defer to another issue. WDYT?

I’d much prefer to keep of `oci_transport.go` consistent. Otherwise there’s no telling which of the alternatives will be updated in the future to make the code consistent, i.e. what use case will break.

An OS-specific check determining the treatment of `:` is OK with me.

mtrmac replied:
Another way to put it, if you _really_ don’t want to update `ParseReference` etc. for Windows, I guess I can live with that, but then you’d have to document the code clearly and explicitly enough to keep your work working in the future.

hferentschik replied:
@mtrmac any chance you mean _ValidatePolicyConfigurationScope_? There seems to be no _ValidatePolicyConfigurationIdentity_.

If you mean _ValidatePolicyConfigurationScope_, what is the expected input to this function. The same as to _ ParseReference_? Something like <dir>:<image>? 

I am still struggling a bit with these policies and scopes. Not quite sure how everything fits together in this area of the code.

hferentschik replied:
@mtrmac another unrelated question. There is one more problem I am seeing which is not addressed in this pull request yet. Could you have a look at https://github.com/containers/image/blob/6e47653e2d514d64af265b69b8cc02542605fccc/copy/copy.go#L290-L290. 

As you know I am exporting a bunch of images from a Docker daemon (running in a VM), store them in an OCI tree on a Windows host and then at some stage re-import into a Docker daemon. This works fine for macOS and Linux. On Windows, I get the error which is referenced above. Effectively: 

    image operating system linux cannot be used on windows

_src.OCIConfig()_ seems to return linux, I would expect Windows though. Any ideas? I have locally commented out the call to _checkImageDestinationForCurrentRuntimeOS_, since I find it questionable either way. Moving forward though, I was wondering how to solve this? If there is no bug in __src.OCIConfig()__ , I was wondering whether it would make sense to add something to SystemContext which allows to skip the check for _ImageDestinationForCurrentRuntimeOS_. WDYT?

mtrmac replied:
> `src.OCIConfig()` seems to return `linux`, I would expect `Windows` though

The image was created by a `dockerd` on Linux, which is what is recorded inside the image’s config. It does not matter what OS is accessing the image or where the image was copied, it is designed to run on Linux.

> I was wondering whether it would make sense to add something to `SystemContext` which allows to skip the check for `ImageDestinationForCurrentRuntimeOS`. WDYT?

Yes; I have implemented that already in https://github.com/containers/image/pull/346/commits/8074c3892f3775ec876a63f8e8e2cd5bda9aea5d for https://github.com/containers/image/pull/346 .

hferentschik replied:
> Yes; I have implemented that already in 8074c38 for #346 .

LOL :-) Seems to be exactly what I need. Except it does not seem to be merged yet :-( 
I will for now focus on the things we discussed so far and leave the destination check for #346. It would be great to get this merged though. Anything I can do do get #346 merged?

mtrmac replied:
> Anything I can do do get #346 merged?

It needs a review by @runcom at least; any extra reviews would be welcome as well.

mtrmac replied:
> @mtrmac any chance you mean _ValidatePolicyConfigurationScope_? There seems to be no _ValidatePolicyConfigurationIdentity_.

Yes, sorry.

> If you mean _ValidatePolicyConfigurationScope_, what is the expected input to this function. The same as to _ ParseReference_? Something like `<dir>:<image>`? 
>
> I am still struggling a bit with these policies and scopes. Not quite sure how everything fits together in this area of the code.

1. References: A way for the user to refer to a _single image_ in text, basically a way to express a `types.ImageReference`. There may be multiple ways to refer to a single underlying image, either by definition (`docker://busybox` vs. `docker://busybox:latest`), or by the environment (`dir:pathtodir` vs. `dir:symlinktodir`). Parsed by `Transport.ParseReference`, formatted by `ImageReference.StringWithinTransport` (i.e. the output of this must be consumable by `ParseReference` and be interpreted as the same underlying image).

2. Policy configuration scopes: A way for the user to define a _hierarchical namespace_ of images. The same underlying image must have exactly one, canonical, “identity” (which is why symlinks are resolved in `NewReference`—but note that the identity fundamentally depends on the environment, i.e. the contents of the symlink), and `PolicyConfigurationNamespaces` returns the namespaces which contain that image (i.e. typically the analogue of the set of its parent directories / paths / hostnames etc.). The values are formatted by `PolicyConfigurationIdentity` and `PolicyConfigurationNamespaces`. There is no method to parse the string back into a Go value (strings in config files are compared for equality with the outputs of these two methods), but there is a `ValidatePolicyConfigurationScope` which should allow the output of `PolicyConfiguration{Identity,Scope}` and preferably little or nothing else, to allow detecting errors in the configuration.

hferentschik commented:
@mtrmac, I am trying to sort out some Windows incompatibilities which stop us to move forward with integrating containers/image into Minishift. See also https://github.com/minishift/minishift/pull/1589.

hferentschik commented:
@runcom 
> I have unfortunately no way to verify this on windows. Changes make sense though

Sure. It would be nice to get more CI system involved and build not only on Linux, but also on other OS types. Minsihift uses CircleCI and AppVeyor for building/testing on macOS resp. Windows. 

For now it might suffice that we are testing this change as part of our integration of containers/image into Minishift - https://github.com/minishift/minishift/pull/1589. 

It would be awesome to get this in, so that we can go ahead on our side as well.

mtrmac commented on oci/layout/oci_dest.go:
```diff
@@ -133,8 +133,12 @@ func (d *ociImageDestination) PutBlob(stream io.Reader, inputInfo types.BlobInfo
+
+	// Chmod is not supported on Windows
+	if runtime.GOOS != "windows" {
+		if err := blobFile.Chmod(0644); err != nil {
```
How does `Chmod` fail here?  Reading `go/src/os/file_posix.go` and `go/src/syscall/syscall_windows.go` it seems to be translated to `GetFileAttributes`+`SetFileAttributes` modifying the readonly attribute, which seems plausible.

hferentschik replied:
There is a difference between os.Chmod and os.File.Chmod. Go figure. See also here https://github.com/google/skicka/issues/77. As a workaround, we can also try to use os.Chmod at this point.

mtrmac replied:
Oops, missed that.  The code is fine, please just add a more detailed explanation, something like
```go
// On POSIX systems, blobFile was created with mode 0600, so we need to make it readable.
// On Windows, the “permissions of newly created files” argument to syscall.Open is
// ignored and the file is already readable; besides, blobFile.Chmod, i.e. syscall.Fchmod,
// always fails on Windows.
```

mtrmac commented on oci/layout/oci_dest.go:
```diff
@@ -144,6 +148,9 @@ func (d *ociImageDestination) PutBlob(stream io.Reader, inputInfo types.BlobInfo
 	}
+
+	// need to explicitly close the file, since a rename won't otherwise not work on Windows
+	blobFile.Close()
```
Per https://golang.org/pkg/io/#Closer

> The behavior of Close after the first call is undefined. 

i.e. the `defer … blobFile.Close()` must be deactivated, similarly to the `succeeded` boolean.

hferentschik replied:
hmm, one still would like defer to kick in until I can explicitly close. I'll take a look.

hferentschik commented:
@mtrmac I pushed an update for everything except _oci_transport.go_. I will look into the latter tomorrow.

mtrmac commented:
> @mtrmac I pushed an update for everything except _oci_transport.go_. 

Thanks, looks good.

hferentschik commented:
@mtrmac I rebased the pull request and changed _ParseReference_ as well as _ValidatePolicyConfigurationScope_ to handle Windows as well. I was tempted to refactor further, but decided against, since I did not want to change any semantics/tests for the non-Windows case. WDYT?

hferentschik commented on copy/copy.go:
```diff
@@ -158,9 +158,12 @@ func Image(policyContext *signature.PolicyContext, destRef, srcRef types.ImageRe
-	if err := checkImageDestinationForCurrentRuntimeOS(src, dest); err != nil {
-		return err
-	}
+	// Disabling checkImageDestinationForCurrentRuntimeOS since the implementation is basing decision on whether there is a match on the current OS
```
@mtrmac I commented this check out. I think the current implementation is not correct and w/o the ability to control this behaviour in some way, I don't think it is really useful. 

I don't think that the OS of the host can be used to determine whether a copy can occur. The destination of the copy might be remote. Or am I missing something?

mtrmac replied:
`dest` _already_ can choose whether this check make sense, via `ImageDestination.MustMatchRuntimeOS`; and this check is actually useful for copying to local `docker-daemon:`.

Maybe a specific transport should change its value of `MustMatchRuntimeOS`? Perhaps `docker-daemon:` _when connecting to a remote host_?

hferentschik replied:
> dest already can choose whether this check make sense, 

Right, but it is hardcoded. In the case of the Docker daemon  _daemon_dest.go_. Given that a Docker daemon can also be remote, having _MustMatchRuntimeOS()_ always return true is preventing me to copy an image from a Windows host to a Docker daemon, even though it should be possible.

> and this check is actually useful for copying to local docker-daemon:.

Maybe. 

> Maybe a specific transport should change its value of MustMatchRuntimeOS? Perhaps docker-daemon: when connecting to a remote host?

Right. So one solution would be to me _daemonImageDestination.MustMatchRuntimeOS()_ stateful. 
In the simplest case one could just use _dockerclient.DefaultDockerHost_. If the default is used, we return _true_, if the SystemContext explicitly specifies a value for _DockerDaemonHost_ we return false (unless it maybe matches `dockerclient.DefaultDockerHost`). What I want to avoid is to make any decisions based on the IP/host specified in _ctx.DockerDaemonHost_. Maybe something like 127.0.0.1 or localhost would still make sense, but everything else is not easily verifiable.  

WDYT? Make _daemonImageDestination.MustMatchRuntimeOS()_ stateful?

hferentschik replied:
@mtrmac Any feedback regarding the other changes in this pull request? Are we good on these?

hferentschik replied:
@mtrmac I made _MustMatchRuntimeOS()_ conditional. WDYT?

mtrmac replied:
> @mtrmac I made MustMatchRuntimeOS() conditional. WDYT?

Looks good. As you say, some of the cases are difficult to determine, and most users shouldn’t hit them. If someone does hits them,. we might need to make the tests a bit more precise and change the semantics; hopefully that won’t affect any previously working users.

hferentschik commented:
@mtrmac, @runcom What do you think about this current state of the pull request?

mtrmac commented on oci/oci.go:
```diff
@@ -1 +1,91 @@
+	if len(groups) != 3 {
+		return reference, nil
+	}
+	return groups[1], &groups[2]
```
In non-Windows it is possible to use relative paths, e.g. `oci:somedir:imagename`. This seems to return a non-`nil` image only if the input is an absolute path; is that intentional?

AFAICT relative paths could be allowed without ambiguity:
- `c:\dir\…` cannot mean “the relative path `.\c:\dir`” because `c:` is an invalid directory name
- `c:\dir…` cannot mean “the image `\dir…` in the relative path `c`” because `\dir` is an invalid image name.
- The troublesome case is `c:dir`, which could mean “the relative path `dir` on the current directory of `c:`” or “the image `dir` in the relative path `c`”, but at least https://blogs.msdn.microsoft.com/oldnewthing/20101011-00/?p=12563 says this concept does not exist in Windows anymore.

The above could be non-blocking, left to someone in the future to improve, because the Windows support is new; more troublesome is that this parses `somedir:busybox:latest` as path `somedir:busybox:latest`, missing image, which would then fail validation in `NewReference`; and, worse, `somedir:busybox` as path `somedir:busybox`, no image, which would succeed but be inconsistent with the POSIX version.

mtrmac replied:
This is still outstanding I think. Or are you fine with `oci:somedir:busybox` just failing?

mtrmac replied:
> and, worse, `somedir:busybox` as path `somedir:busybox`, no image, which would succeed but be inconsistent with the POSIX version.

Sorry, that would pass validation but `somedir:busybox` is an invalid path on Windows, so it would fail either way.

So, overall, with this PR relative paths with images just fail; that’s clearly not worse than before this PR, so non-blocking for me.

mtrmac commented on oci/oci.go:
```diff
@@ -1 +1,91 @@
+
+	// Refuse also "/", otherwise "/" and "" would have the same semantics,
+	// and "" could be unexpectedly shadowed by the "/" entry.
+	// (Note: we do allow "/:someimage", a bit ridiculous but why refuse it?)
```
The parenthesized note is no longer relevant now that identities/scopes do not contain the image name.

hferentschik replied:
I'll remove it

mtrmac commented on oci/oci.go:
```diff
@@ -1 +1,91 @@
+	if runtime.GOOS == "windows" {
+		return validateScopeWindows(scope, dir)
+	}
+	return validateScopeNonWindows(scope, dir)
```
Can the two variants share code via `filepath.IsAbs`? (That one also handles UNC paths, but that seems to be compatible enough with our formats.)

hferentschik replied:
Potentially. I guess you are talking about just calling "filepath.IsAbs" w/o a OS specific switch. There would still be 

    if scope == "/" {
        return errors.New(`Invalid scope "/": Use the generic default scope ""`)
    }

to take care of which is non Windows specific (unless you suggest we drop this).

Another disadvantage of using filepath.IsAbs is that one cannot easily write tests for Windows paths. In the current code I can test _validateScopeWindows_ even w/o running the test suite on Windows. Once filepath.IsAbs is used that is not the case anymore. 

I am undecided. I see pro and cons for both and tempted to prefer the current code to start with.

mtrmac replied:
Sure, the non-Windows testing is a valid concern. I suppose we don’t need to care about UNC paths at least for now.

mtrmac commented on oci/oci_test.go:
```diff
@@ -0,0 +1,71 @@
+		{`D:\:apline`, `D:\`, ""},
+		{"C:", "C:", "Invalid scope 'C:'. Must be an absolute path"},
+		{"E", "E", "Invalid scope 'E'. Must be an absolute path"},
+		{"", "", "Invalid scope ''. Must be an absolute path"},
```
(`ValidatePolicyConfigurationScope` is never called with the `""` value, so this case is unnecessary. Doesn’t hurt though.)

hferentschik replied:
ok

mtrmac commented on oci/layout/oci_transport.go:
```diff
@@ -111,12 +90,19 @@ func NewReference(dir, image string) (types.ImageReference, error) {
+	} else {
+		if strings.Contains(resolved, ":") {
+			return nil, errors.Errorf("Invalid OCI reference %s:%s: path %s contains a colon", dir, image, resolved)
+		}
```
One more thing, prompted by https://github.com/projectatomic/skopeo/issues/455 :

Nowadays we don’t treat `:` in policy scopes specially — but we _do_ treat `:` specially in the non-canonicalized input path, as output in `StringWithinTransport` and re-parsed in `ParseReference`. So, actually, this check for colons should, if I am not mistaken again, validate `dir`, not `resolved`.

(We originally didn’t validate `dir` because `image` was not allowed to contain colons, so `dir_with_colons:image_without_colons` was not ambiguous. After `image` was allowed colons, this became ambiguous and we should have started refusing such `dir` values at that time.

Conceptually independently, but at at about the same time, `image` was dropped from `PolicyConfiguration*`, and the `resolved` check was no longer necessary.)

hferentschik replied:
ahh, so check is against the unresolved path. ok

mtrmac commented on oci/layout/oci_transport.go:
```diff
@@ -36,40 +37,21 @@ func (t ociTransport) ParseReference(reference string) (types.ImageReference, er
-		}
+	dir, image := oci.SplitDirAndImage(scope)
+	if image != nil && !oci.ValidImageName(*image) {
+		return errors.Errorf("Invalid image %s", image)
```
… see the other new comment; `NewReference` maybe does not need refuse colons in `resolved` at all; in which case this should do just the check for absolute paths which are not `/`.

hferentschik replied:
Ahh, that would make things easier indeed. Needs a bit of refactoring ....

hferentschik commented on oci/oci.go:
```diff
@@ -1 +1,91 @@
+// The image is returned as string pointer, allowing for nil being returned, if no image part could be extraced from the reference.
+// Neither directory nor image parts are validated at this stage.
+func SplitDirAndImage(reference string) (string, *string) {
+	if runtime.GOOS == "windows" {
```
>  OTOH a shared implementation ensures our CI is also testing most of the Windows code, so this is non-blocking.

Right, that's the intention. It allows that the core expectations are tests even on the current CI system. Using platform specific files loses this. If there were a Windows-based CI (like AppVeyor) setup, I'd prefer the platform specific files, but w/o I rather have some tests which are exercised all the time.

mtrmac replied:
Makes sense.

hferentschik commented on oci/oci.go:
```diff
@@ -1 +1,91 @@
+// SplitDirAndImage tries to split the provided OCI reference into the OCI directory and image.
+// The image is returned as string pointer, allowing for nil being returned, if no image part could be extraced from the reference.
+// Neither directory nor image parts are validated at this stage.
+func SplitDirAndImage(reference string) (string, *string) {
```
+1

hferentschik commented on oci/layout/oci_transport.go:
```diff
@@ -92,14 +74,11 @@ type ociReference struct {
+		return NewReference(dir, "")
 	}
-	return NewReference(dir, image)
+	return NewReference(dir, *image)
```
I know. But as said, I wanted to keep the existing code behaving exactly how it was. And for that I need to differentiate between "" and nil. If you are ok to return "" and that an existing test gets changed, then yes.

mtrmac replied:
> I know. But as said, I wanted to keep the existing code behaving exactly how it was. And for that I need to differentiate between "" and nil.

All `ParseReference` does now is call `SplitDirAndImage`, and replace a `nil` with `""`. The difference is immediately erased.

Or are you talking (with the references to “existing test” about `ValidatePolicyConfigurationScope`)? That one does treat a trailing `:` and no `:` differently — but `ValidatePolicyConfigurationScope` should nowadays not expect a `:image` in the input at all, so that will disappear.

hferentschik commented on oci/archive/oci_transport.go:
```diff
@@ -48,52 +49,30 @@ func (t ociArchiveTransport) ParseReference(reference string) (types.ImageRefere
-
-	if len(sep) == 2 {
-		image = sep[1]
+	dir, image := oci.SplitDirAndImage(reference)
```
ok

hferentschik commented on oci/archive/oci_transport.go:
```diff
@@ -102,12 +81,14 @@ func NewReference(file, image string) (types.ImageReference, error) {
 	// from being ambiguous with values of PolicyConfigurationIdentity.
 	if strings.Contains(resolved, ":") {
 		return nil, errors.Errorf("Invalid OCI reference %s:%s: path %s contains a colon", file, image, resolved)
 	}
```
ok

hferentschik commented:
@mtrmac, I pushed an update to the pull request incorporating your feedback. Hopefully, I understood all your comments correctly. Turns out things are getting easier and the code looks more regular which is usually a good sign. WDYT?

mtrmac commented on oci/archive/oci_transport.go:
```diff
@@ -48,51 +48,16 @@ func (t ociArchiveTransport) ParseReference(reference string) (types.ImageRefere
+	if err != nil {
+		return err
 	}
 	return nil
```
(This could be just `return internal.ValidateScope(scope)`.)

hferentschik replied:
right.

mtrmac commented on oci/internal/oci_util.go:
```diff
@@ -0,0 +1,121 @@
+// SplitDirAndImage tries to split the provided OCI reference into the OCI directory and image.
+// The image is returned as string pointer, allowing for nil being returned, if no image part could be extraced from the reference.
+// Neither directory nor image parts are validated at this stage.
+func SplitDirAndImage(reference string) (string, string) {
```
Please rename this `dir` to `path`, throughout.

The comment about a string pointer is no longer true.

hferentschik replied:
+1

mtrmac commented on oci/internal/oci_util.go:
```diff
@@ -0,0 +1,121 @@
+	return nil
+}
+
+// ValidateScope takes the OCI directory and validates its scope.
```
“validates a policy configuration scope for an OCI transport” or something like that; for `oci-archive` this can be just an absolute path to a file.

hferentschik replied:
+1

mtrmac commented on oci/archive/oci_transport.go:
```diff
@@ -102,12 +67,13 @@ func NewReference(file, image string) (types.ImageReference, error) {
-		return nil, errors.Errorf("Invalid OCI reference %s:%s: path %s contains a colon", file, image, resolved)
+
+	err = internal.ValidateOCIPath(file)
+	if err != nil {
```
(This could be `if err := … err != nil {`.  Applies to both transports.)

hferentschik replied:
sure. Personally, I don't like this style of the if statement, but fair enough

hferentschik commented:
@mtrmac pushed an update. 

> The question of accepting relative paths is still outstanding; but if you prefer merging the PR now without making that possible, that’s fine with me, as long as it is a conscious decision.

I'd like to get this merged, since this allows Minishift to continue.

hferentschik commented:
@runcom, @mtrmac - what do you think, can we merge this now?

mtrmac commented:
:+1:

> > The question of accepting relative paths is still outstanding; but if you prefer merging the PR now without making that possible, that’s fine with me, as long as it is a conscious decision.
>
> I'd like to get this merged, since this allows Minishift to continue. 

OK. Thanks!

@runcom PTAL.

hferentschik commented:
@runcom not to seem impatient, but I have two pull requests on the Minishift side depending on this :-) :

* https://github.com/minishift/minishift/pull/1685
* https://github.com/minishift/minishift/pull/1716

runcom commented:
lgtm

hferentschik commented:
thanks :-)
